### DISCUSSION TITLE
Trailing Commas for nested/sub-arrays

### DISCUSSION THREAD

**Discussion by arwelcmm** [squizlabs/PHP_CodeSniffer#3850] [Q&A]
Hi guys, is there a rule that appends a trailing comma to sub-arrays/nested arrays? I have implemented the following phpcs.xml in my current project, but it was flagged by a colleague that the current xml does not append trailing commas to sub-arrays, but works fine for non-nested arrays. If anyone has any insight I'd be very grateful!

    <rule ref="PSR12"/>

    
    
    <rule ref="Generic.Files.LineLength">
        <properties>
            <property name="lineLimit" value="175"/>
            <property name="absoluteLineLimit" value="220"/>
        </properties>
    </rule>
    <rule ref="PSR12"></rule>
    <rule ref="Squiz.Arrays.ArrayDeclaration">
        <exclude name="Squiz.Arrays.ArrayDeclaration.ValueNotAligned" />
        <exclude name="Squiz.Arrays.ArrayDeclaration.KeyNotAligned" />
        <exclude name="Squiz.Arrays.ArrayDeclaration.CloseBraceNotAligned" />
        <exclude name="Squiz.Arrays.ArrayDeclaration.MultiLineNotAllowed" />
        <exclude name="Squiz.Arrays.ArrayDeclaration.SingleLineNotAllowed" />
        <exclude name="Squiz.Arrays.ArrayDeclaration.DoubleArrowNotAligned" />
        <exclude name="Squiz.Arrays.ArrayDeclaration.KeySpecified" />
    </rule>
    <file>config/</file>
    <file>src/</file>
    <file>tests/</file>

**Comment by fredden** [comment]
@arwelcmm can you provide a code snippet which demonstrates the problem you're trying to solve? Perhaps a 'bad' and 'fixed' sample with both 'non-nested' and 'sub-arrays'? I think I know what you're talking about, but I want to make sure. Also, having a code snippet will make it much easier to test if a sniff behaves as you want/expect.

**Reply by arwelcmm** [reply]
Cheers for the quick response @fredden. Here's a simple snippet that illustrates the issue, this is following a run of phpcbf. Hope this is helpful!

![image](https://github.com/squizlabs/PHP_CodeSniffer/assets/135008541/54bd90b1-db5b-445a-a741-670fb187f5e5)

You'll notice phpcbf has put the trailing commas at the end of of the arrays nested in once, but has missed one that's nested in twice (on line 22)

**Reply by jrfnl** [reply]
@arwelcmm Kind request to post the actual code and not a screenshot. We cannot test anything with a screenshot ...

**Reply by fredden** [reply]
When I run `bin/phpcbf --basepath=. -ps --standard=Squiz --sniffs=Squiz.Arrays.ArrayDeclaration one.php two.php` on the following code snippets, they both get all the extra commas inserted automatically. I don't see anywhere that lacks a trailing comma. What version of PHP_CodeSniffer are you using?

<details><summary><code>one.php</code></summary>

```php
<?php

return [
    'current' => [
        'single' => [
            'credits' => 1,
            'value' => 8
        ]
    ]
];
```

</details><details><summary><code>two.php</code></summary>

```php
<?php

return [
    'current' => [
        'single' => [
            'credits' => 1,
            'value' => 8
        ],
        'both' => [
            'credits' => 1,
            'value' => 8
        ]
    ],
    'historical' => [
        '2023-02-27' => [
            'single' => 6,
            'both' => 6
        ],
        '2020-01-01' => [
            'single' => 5,
            'both' => 7.50
        ]
    ]
];
```

</details>

**Reply by jrfnl** [reply]
I suspect the problem is not the sniff in and of itself, but the exclusions in the ruleset. That particular sniff is problematic/unwieldy, in that exclusions will often also exclude multiple other errorcodes in the code flow after the code which handles that particular error.

Also see https://github.com/squizlabs/PHP_CodeSniffer/issues/582

**Reply by jrfnl** [reply]
Not a terribly clean solution to your problem, but one which should work all the same:
Instead of excluding those error codes, set the error `severity` for each of those error codes to `1`. As PHPCS by default only reports on severity `5` and up, this should allow the code to continue and reach the trailing comma related code checks.

That way, the other errors won't show unless you set `--severity=1`.

```xml
	<rule ref="Squiz.Arrays.ArrayDeclaration.ValueNotAligned">
		<severity>1</severity>
	</rule>
```

**Comment by arwelcmm** [comment]
> @arwelcmm Kind request to post the actual code and not a screenshot. We cannot test anything with a screenshot ...

Of course, my apologies! Will endeavour to do this going forward, a bit daft of me in hindsight...

**Comment by arwelcmm** [comment]
> When I run `bin/phpcbf --basepath=. -ps --standard=Squiz --sniffs=Squiz.Arrays.ArrayDeclaration one.php two.php` on the following code snippets, they both get all the extra commas inserted automatically. I don't see anywhere that lacks a trailing comma. What version of PHP_CodeSniffer are you using?
> 
> `one.php`
> ```
> <?php
> 
> return [
>     'current' => [
>         'single' => [
>             'credits' => 1,
>             'value' => 8
>         ]
>     ]
> ];
> ```
> 
> `two.php`
> ```
> <?php
> 
> return [
>     'current' => [
>         'single' => [
>             'credits' => 1,
>             'value' => 8
>         ],
>         'both' => [
>             'credits' => 1,
>             'value' => 8
>         ]
>     ],
>     'historical' => [
>         '2023-02-27' => [
>             'single' => 6,
>             'both' => 6
>         ],
>         '2020-01-01' => [
>             'single' => 5,
>             'both' => 7.50
>         ]
>     ]
> ];
> ```
I'm using PHP_Codesniffer version 3.7.2.

**Comment by arwelcmm** [comment]
> Not a terribly clean solution to your problem, but one which should work all the same: Instead of excluding those error codes, set the error `severity` for each of those error codes to `1`. As PHPCS by default only reports on severity `5` and up, this should allow the code to continue and reach the trailing comma related code checks.
> 
> That way, the other errors won't show unless you set `--severity=1`.
> 
> ```
> 	<rule ref="Squiz.Arrays.ArrayDeclaration.ValueNotAligned">
> 		<severity>1</severity>
> 	</rule>
> ```
Thanks for the suggestion! Unfortunately changing my xml file to match this hasn't solved it. I've spoken to the colleague who raised the issue and he's clarified that it only seems to be happening on the final index of a nested array. Hope this extra information might help!

**Reply by jrfnl** [reply]
@arwelcmm In that case you may need to file a bug report.

**Comment by arwelcmm** [comment]
> @arwelcmm In that case you may need to file a bug report.
No problem. I shall get one open now. Thanks for all your help!
